### PR TITLE
Closes #124: Add menu items to visualizations with export to excel as…

### LINKED ISSUE #124: Add Export to Excel to Data Table Visualizations

### PR DESCRIPTION
**Pull request by pml984**
… the initial item.

### CODE REVIEW

File: src/js/components/visualization/ChartComponent.jsx [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import * as Components from 'safe-framework'
 import React, {Component, PropTypes} from 'react'
 
+/* global document */
```
Comment (jeffshaver) [inline src/js/components/visualization/ChartComponent.jsx]: shouldn't be here. this is a c9 thing

File: src/js/components/visualization/ChartComponent.jsx [hunk 2]
```diff
@@ -83,15 +85,13 @@ export class ChartComponent extends Component {
         data={chartData}
         drilldown={drillDownFieldName != null}
         options={options}
-        ref='chart'
+        ref={(ref) => (this.chart = ref)}
```
Comment (jeffshaver) [inline src/js/components/visualization/ChartComponent.jsx:92]: is there a reason to not just just `ref='chart'` and then reference it as `this.refs.chart`? is this another weird ref issue with react?

File: src/js/components/visualization/ChartMenu.jsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import changeCase from 'change-case'
+import Download from 'material-ui/svg-icons/file/file-download'
+import MenuItem from 'material-ui/MenuItem/MenuItem'
+import {saveCanvas} from '../../modules/utilities'
+import React, {Component, PropTypes} from 'react'
+
+export class ChartMenu extends Component {
```
Comment (jeffshaver) [inline src/js/components/visualization/ChartMenu.jsx:7]: This should be a functional component since it doesn't use any lifecycle methods

File: src/js/components/visualization/MapMenu.jsx [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import {Component, PropTypes} from 'react'
```
Comment (jeffshaver) [inline src/js/components/visualization/MapMenu.jsx:1]: should be a functional component. however, why does this just return nothing?

File: src/js/components/visualization/TableMenu.jsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import Download from 'material-ui/svg-icons/file/file-download'
```
Comment (jeffshaver) [inline src/js/components/visualization/TableMenu.jsx:3]: should be a functional component

File: src/js/components/visualization/Visualization.jsx [hunk 6]
```diff
@@ -57,13 +60,25 @@ class Visualization extends Component {
...
-
+    const VisualizationMenu = menus[`${visualizationComponentName}Menu`]
+      
     return (
       <Paper style={styles.paper}>
-        <VisualizationToolbar title={name} />
+        <VisualizationToolbar
+          title={name}
+        >
+          {VisualizationMenu
```
Comment (jeffshaver) [inline src/js/components/visualization/Visualization.jsx:72]: I would prefer if these brackets were aligned. i.e. on their own separate lines. helps figure out where it ends

File: src/js/components/visualization/Visualization.jsx [hunk 7]
```diff
@@ -57,13 +60,25 @@ class Visualization extends Component {
...
+              metadata={visualization}
+              visualization={this}
+            />
+            : null}
+        </VisualizationToolbar>
         <VisualizationComponent
           data={data}
           metadata={visualization}
+          name={name}
+          ref={(ref) => (this.component = ref)}
```
Comment (jeffshaver) [inline src/js/components/visualization/Visualization.jsx]: similar comment to one above. is there a reason to do this as a function?

File: src/js/components/visualization/VisualizationToolbar.jsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    const {children, title} = this.props
+    
+    return (
+      <Toolbar>
+        <ToolbarGroup float='left'>
+          <ToolbarTitle text={title} />
+        </ToolbarGroup>
+        <ToolbarGroup float='right'>
+          {/* Icons to float to right */}
+          {children
```
Comment (jeffshaver) [inline src/js/components/visualization/VisualizationToolbar.jsx:30]: similar to above. align the brackets by putting them onto new lines

File: src/js/modules/utilities.js [hunk 9]
```diff
@@ -1,3 +1,7 @@
+import changeCase from 'change-case'
+
+/* global document */
```
Comment (jeffshaver) [inline src/js/modules/utilities.js:3]: remove

File: src/js/modules/utilities.js [hunk 10]
```diff
@@ -22,6 +26,17 @@ export const excludeEmptyFilters = (filters) => (
   })
 )
 
+export const saveCanvas = (canvas, name) => {
+  const image = canvas.toDataURL()
+  const aLink = document.createElement('a')
+  const evt = document.createEvent('HTMLEvents')
+  
+  evt.initEvent('click')
+  aLink.download = changeCase.pascalCase(name) + '.png'
```
Comment (jeffshaver) [inline src/js/modules/utilities.js]:
```
aLink.download = `${changeCase.pascalCase(name)}.png`
```

File: src/js/components/visualization/TableMenu.jsx [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  render () {
+    const {visualization} = this.props
+  
+    return (
+      <span>
+        <MenuItem
+          key='exportExcel'
+          leftIcon={<Download />}
+          primaryText='Export'
+          onTouchTap={function (event) {
```
Comment (jeffshaver) [inline src/js/components/visualization/TableMenu.jsx]: don't define callbacks inline. when you do, they get created on every render call. if we find this elsewhere in the code, it should be re-worked if possible

File: src/js/components/visualization/TableComponent.jsx [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        field
+      }))
+  
+    return (
+      <DataTable
+        columns={columns}
+        data={data}
+        enableColResize='true'
+        enableSorting='true'
+        exportFileName={changeCase.pascalCase(name) + '.csv'}
```
Comment (jeffshaver) [inline src/js/components/visualization/TableComponent.jsx]:
```
exportFileName={`${changeCase.pascalCase(name).csv`}
```

File: src/js/components/visualization/TableComponent.jsx [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    type: PropTypes.string.isRequired
+  }
+
+  static getToolbarItems (visualization) {
+    return [
+      <MenuItem
+        key='exportExcel'
+        leftIcon={<Download />}
+        primaryText='Export'
+        onTouchTap={function (event) {
```
Comment (jeffshaver) [inline src/js/components/visualization/TableComponent.jsx]: don't define callbacks inline. when you do, they get created on every render call. if we find this elsewhere in the code, it should be re-worked if possible

File: src/js/components/visualization/ChartMenu.jsx [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+  render () {
+    const {metadata, visualization} = this.props
+    
+    return (
+      <span>
+        <MenuItem
+          key='saveChart'
+          leftIcon={<Download />}
+          primaryText='Save'
+          onTouchTap={function (event) {
```
Comment (jeffshaver) [inline src/js/components/visualization/ChartMenu.jsx]: don't define callbacks inline. when you do, they get created on every render call. if we find this elsewhere in the code, it should be re-worked if possible

File: src/js/components/visualization/MapMenu.jsx (hunk 4 above)
Comment (pml984) [inline src/js/components/visualization/MapMenu.jsx:1]: Just a placeholder right now until we get any menu items.

File: src/js/components/visualization/ChartComponent.jsx (hunk 2 above)
Comment (pml984) [inline src/js/components/visualization/ChartComponent.jsx:92]: I saw this in the React documentation. I thought it was a little cleaner instead of having to reference the 'refs' object each time to get a reference. I don't mind either way though. We could also use underscores to ensure that there are not confliccts. We should probably figure out which way we prefer and stick with it to be consistent

File: src/js/components/visualization/TableComponent.jsx [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+      }))
+  
+    return (
+      <DataTable
+        columns={columns}
+        data={data}
+        enableColResize='true'
+        enableSorting='true'
+        exportFileName={changeCase.pascalCase(name) + '.csv'}
+        ref={(ref) => (this.table = ref)}
```
Comment (jeffshaver) [inline src/js/components/visualization/TableComponent.jsx:29]: actually, why is this here? this.table isn't ever used?

File: src/js/components/visualization/Visualization.jsx (hunk 7 above)
Comment (jeffshaver) [inline src/js/components/visualization/Visualization.jsx]: this.component is never used

File: src/js/components/visualization/ChartComponent.jsx (hunk 2 above)
Comment (jeffshaver) [inline src/js/components/visualization/ChartComponent.jsx:92]: I am fine with doing `ref={(ref) => (this._chart = ref)}`. so callbacks and underscores.
